### ISSUE TITLE
Support for Ascend NPU hardware

### ISSUE THREAD

**Issue by JingWoo** [ollama/ollama#5315]
Huawei Ascend AI processor is an AI chip based on Huawei-developed Da Vinci architecture. It performs well in processing large-scale data and complex computing tasks. Currently, the llama.cpp project is adapting to the Ascend series AI processors. I'm also adapting Ollama to support the Ascend series of AI processors to expand the hardware ecosystem Ollama supports.

**Comment by mchiang0610** [comment]
Thanks @JingWoo for reaching out! Do you happen to know the Huawei team? We test on devices before adding support. For most of the integrations, we spend the time directly with hardware vendors to test out their drivers prior to a release.

**Comment by JingWoo** [comment]
Yes, I'm from Huawei and I now trying adding support Ascend for ollama. 
I also have some Ascend AI servers in hand, so adding support and testing on them.
If you have some more problems/features needed on Ascend, please feel free to contact with me.

**Comment by JingWoo** [comment]
@mchiang0610 Yes, I'm from Huawei and I now trying adding support Ascend for ollama. 
I also have some Ascend AI servers in hand, so adding support and testing on them.
If you have some more problems/features needed on Ascend, please feel free to contact with me.

**Comment by mchiang0610** [comment]
@JingWoo Would it be possible to see if we can add an Ascend AI server for testing? Where could we request hardware for testing? 

Every single Ollama release goes through testing before release. We've been working with hardware vendors to build out a test matrix for stability and reliability checks.

**Comment by cosren** [comment]
@JingWoo  you can  setup ollama on the acend npu divice ? Do you have any guide doc for it ?

**Comment by myoss** [comment]
hope it can support running on Huawei Ascend servers

**Comment by JingWoo** [comment]
> @JingWoo you can setup ollama on the acend npu divice ? Do you have any guide doc for it ?

Yes, we have currently adapted Ollama to support Ascend devices，please refer to https://github.com/ollama/ollama/pull/5872

**Comment by JingWoo** [comment]
demo:

**Comment by zhongTao99** [comment]
We have completed the adaptation of Ascend and verified that ollama uses Ascend NPU devices.
![origin](https://github.com/user-attachments/assets/04031a77-dfe8-47b8-9b58-e9d262c74216)

**Comment by JerryYao80** [comment]
@zhongTao99 请教一个问题，在容器内外，是不是310P的CardID会改变？
在容器外：
+--------------------------------------------------------------------------------------------------------+
| npu-smi 24.1.rc2                                 Version: 24.1.rc2                                     |
+-------------------------------+-----------------+------------------------------------------------------+
| NPU     Name                  | Health          | Power(W)     Temp(C)           Hugepages-Usage(page) |
| Chip    Device                | Bus-Id          | AICore(%)    Memory-Usage(MB)                        |
+===============================+=================+======================================================+
| 1       310P3                 | OK              | NA           59                0     / 0             |
| 0       0                     | 0000:01:00.0    | 0            1397 / 44280                            |
+-------------------------------+-----------------+------------------------------------------------------+
| 1       310P3                 | OK              | NA           58                0     / 0             |
| 1       1                     | 0000:01:00.0    | 0            1520 / 43693                            |
+===============================+=================+======================================================+
| 2       310P3                 | OK              | NA           58                0     / 0             |
| 0       2                     | 0000:02:00.0    | 0            1599 / 44280                            |
+-------------------------------+-----------------+------------------------------------------------------+
| 2       310P3                 | OK              | NA           56                0     / 0             |
| 1       3                     | 0000:02:00.0    | 0            1264 / 43693                            |
+===============================+=================+======================================================+
| 4       310P3                 | OK              | NA           54                0     / 0             |
| 0       4                     | 0000:81:00.0    | 0            1254 / 44280                            |
+-------------------------------+-----------------+------------------------------------------------------+
| 4       310P3                 | OK              | NA           55                0     / 0             |
| 1       5                     | 0000:81:00.0    | 0            1612 / 43693                            |
+===============================+=================+======================================================+
| 5       310P3                 | OK              | NA           53                0     / 0             |
| 0       6                     | 0000:82:00.0    | 0            1355 / 44280                            |
+-------------------------------+-----------------+------------------------------------------------------+
| 5       310P3                 | OK              | NA           51                0     / 0             |
| 1       7                     | 0000:82:00.0    | 0            1508 / 43693                            |
+===============================+=================+======================================================+

在容器内就变成了：
+-------------------------------+-----------------+------------------------------------------------------+
| NPU     Name                  | Health          | Power(W)     Temp(C)           Hugepages-Usage(page) |
| Chip    Device                | Bus-Id          | AICore(%)    Memory-Usage(MB)                        |
+===============================+=================+======================================================+
| 0       310P3                 | OK              | NA           59                0     / 0             |
| 0       0                     | 0000:01:00.0    | 0            1397 / 44280                            |
+-------------------------------+-----------------+------------------------------------------------------+
| 0       310P3                 | OK              | NA           58                0     / 0             |
| 1       1                     | 0000:01:00.0    | 0            1520 / 43693                            |
+===============================+=================+======================================================+
| 32      310P3                 | OK              | NA           58                0     / 0             |
| 0       2                     | 0000:02:00.0    | 0            1600 / 44280                            |
+-------------------------------+-----------------+------------------------------------------------------+
| 32      310P3                 | OK              | NA           56                0     / 0             |
| 1       3                     | 0000:02:00.0    | 0            1264 / 43693                            |
+===============================+=================+======================================================+
| 32768   310P3                 | OK              | NA           54                0     / 0             |
| 0       4                     | 0000:81:00.0    | 0            1255 / 44280                            |
+-------------------------------+-----------------+------------------------------------------------------+
| 32768   310P3                 | OK              | NA           55                0     / 0             |
| 1       5                     | 0000:81:00.0    | 0            1611 / 43693                            |
+===============================+=================+======================================================+
| 32800   310P3                 | OK              | NA           52                0     / 0             |
| 0       6                     | 0000:82:00.0    | 0            1355 / 44280                            |
+-------------------------------+-----------------+------------------------------------------------------+
| 32800   310P3                 | OK              | NA           51                0     / 0             |
| 1       7                     | 0000:82:00.0    | 0            1507 / 43693                            |
+===============================+=================+======================================================+

**Comment by geweixuan** [comment]
> @myoss @cosren We have completed the adaptation of Ascend and verified that ollama uses Ascend NPU devices. ![origin](https://private-user-images.githubusercontent.com/56594937/352907391-04031a77-dfe8-47b8-9b58-e9d262c74216.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMDgwNDUsIm5iZiI6MTczMDEwNzc0NSwicGF0aCI6Ii81NjU5NDkzNy8zNTI5MDczOTEtMDQwMzFhNzctZGZlOC00N2I4LTliNTgtZTlkMjYyYzc0MjE2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI4VDA5MjkwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwZDllMjY5MWFkNGNiYjI1YjljMTM0ZWU4ZmZhZmZhZWJlNDFlMzk3NzRkMzUwNzJmM2VhN2E2MjU5YTgxZGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mMWdkqIXNAV2PeTtYryWS52zOVAZxheOXvO2V06fEO4) [ ![origin](https://private-user-images.githubusercontent.com/56594937/352907391-04031a77-dfe8-47b8-9b58-e9d262c74216.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMDgwNDUsIm5iZiI6MTczMDEwNzc0NSwicGF0aCI6Ii81NjU5NDkzNy8zNTI5MDczOTEtMDQwMzFhNzctZGZlOC00N2I4LTliNTgtZTlkMjYyYzc0MjE2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI4VDA5MjkwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwZDllMjY5MWFkNGNiYjI1YjljMTM0ZWU4ZmZhZmZhZWJlNDFlMzk3NzRkMzUwNzJmM2VhN2E2MjU5YTgxZGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mMWdkqIXNAV2PeTtYryWS52zOVAZxheOXvO2V06fEO4) ](https://private-user-images.githubusercontent.com/56594937/352907391-04031a77-dfe8-47b8-9b58-e9d262c74216.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMDgwNDUsIm5iZiI6MTczMDEwNzc0NSwicGF0aCI6Ii81NjU5NDkzNy8zNTI5MDczOTEtMDQwMzFhNzctZGZlOC00N2I4LTliNTgtZTlkMjYyYzc0MjE2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI4VDA5MjkwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwZDllMjY5MWFkNGNiYjI1YjljMTM0ZWU4ZmZhZmZhZWJlNDFlMzk3NzRkMzUwNzJmM2VhN2E2MjU5YTgxZGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mMWdkqIXNAV2PeTtYryWS52zOVAZxheOXvO2V06fEO4) [ ](https://private-user-images.githubusercontent.com/56594937/352907391-04031a77-dfe8-47b8-9b58-e9d262c74216.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMDgwNDUsIm5iZiI6MTczMDEwNzc0NSwicGF0aCI6Ii81NjU5NDkzNy8zNTI5MDczOTEtMDQwMzFhNzctZGZlOC00N2I4LTliNTgtZTlkMjYyYzc0MjE2LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI4VDA5MjkwNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwZDllMjY5MWFkNGNiYjI1YjljMTM0ZWU4ZmZhZmZhZWJlNDFlMzk3NzRkMzUwNzJmM2VhN2E2MjU5YTgxZGQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mMWdkqIXNAV2PeTtYryWS52zOVAZxheOXvO2V06fEO4)

大佬是怎么做到的呢，我也很感兴趣

**Comment by HartleyLau** [comment]
I look forward to this PR [5872](https://github.com/ollama/ollama/pull/5872) being merged，I'm currently using Huawei Ascend NPU hardware

**Comment by James4Ever0** [comment]
> @JingWoo Would it be possible to see if we can add an Ascend AI server for testing? Where could we request hardware for testing? 
> 
> Every single Ollama release goes through testing before release. We've been working with hardware vendors to build out a test matrix for stability and reliability checks. 

Hello from China! What are your specific hardware requirements for the test environment? For example, is remote or cloud environment OK or local only? How many GPU cards are needed? What kind of GPU would be required for testing?

**Comment by wangcong19940121** [comment]
@JingWoo 
您好，请教一个问题。我这边leopony/ollama-cann-300i-duo镜像文件成功运行，启动ollama服务，识别出部署的Ascend310（不是310P）。
 
容器ollama list成功识别到本地模型deepseek-r1。

#
出现问题：
run大模型的时候，出现如下错误：Error:llama runner process has terminated:error load model:vector::_M_range_check:_n(which is 6)>=this->size()(which is 6)。
 
对此不甚理解，是因为Ascend310上没有300i-duo？还是镜像文件的问题？还是适配环境出现的问题？

**Comment by wangcong19940121** [comment]
@JingWoo 
您好，请教一个问题。我这边leopony/ollama-cann-300i-duo镜像文件成功运行，启动ollama服务，识别出部署的Ascend310（不是310P）。
 
容器ollama list成功识别到本地模型deepseek-r1。

#
出现问题：
run大模型的时候，出现如下错误：Error:llama runner process has terminated:error load model:vector::_M_range_check:_n(which is 6)>=this->size()(which is 6)。
 
对此不甚理解，是因为Ascend310上没有300i-duo？还是镜像文件的问题？还是适配环境出现的问题？

**Comment by belmont** [comment]
> Huawei Ascend AI processor is an AI chip based on Huawei-developed Da Vinci architecture. It performs well in processing large-scale data and complex computing tasks. Currently, the llama.cpp project is adapting to the Ascend series AI processors. I'm also adapting Ollama to support the Ascend series of AI processors to expand the hardware ecosystem Ollama supports.

for Atlas 300I Duo please also add FFmpeg acceleration support which is a must for many Computer Vision solutions
